src/wg/format/png/reader.d

```diff
@@ -0,0 +1,926 @@
+            input = input[lineBytes..$];
+        }
+        result.rowPitch = pixelSize * width;
+        return true;
+    }
+
+    private bool moveRgbTransparencyToAlpha() nothrow @nogc
+    {
+        immutable newLineBytes = bitDepth * 4 * width / 8;
+        immutable newTotalBytes = newLineBytes * height;
+        auto output = cast(ubyte[])allocator.allocate(newTotalBytes);
+        if (output.length != newTotalBytes) return withError("Failed allocating pixels data from the given allocator");
+        auto input = pixels;
+        auto toDeallocate = pixels;
+        scope (exit) allocator.deallocate(toDeallocate);
+        pixels = output;
+
+        immutable int samples = width * 3;
+
+        if (bitDepth == 16)
+        {
+            auto input16 = input.asArrayOf!ushort;
+            auto output16 = output.asArrayOf!ushort;
+            immutable transparency16 = transparency.asArrayOf!ushort;
+            foreach (y; 0..height)
+            {
+                int writePos = 0;
+                int x = 0;
+                while (x < samples)
+                {
+                    output16[writePos++] = input16[x];
+                    output16[writePos++] = input16[x+1];
+                    output16[writePos++] = input16[x+2];
+                    if (input16[x..x + 3] == transparency16[0..3]) output16[writePos++] = 0;
+                    else output16[writePos++] = 0xffff;
+                    x += 3;
+                }
+                output16 = output16[writePos..$];
+                input16 = input16[samples..$];
+            }
+        }
+        else
+        {
+            foreach (y; 0..height)
+            {
+                int writePos = 0;
+                int x = 0;
+                while (x < samples)
+                {
+                    output[writePos++] = input[x];
+                    output[writePos++] = input[x+1];
+                    output[writePos++] = input[x+2];
+                    if (input[x..x + 2] == transparency[0..3]) output[writePos++] = 0;
+                    else output[writePos++] = 0xff;
+                    x += 3;
+                }
+                output = output[newLineBytes..$];
+                input = input[lineBytes..$];
+            }
+        }
+
+        result.rowPitch = newLineBytes;
+        return true;
+    }
+
+    private bool setupFormat() nothrow @nogc
+    {
+        static immutable string[] formats = [
+            "",
+            "l_1",
+            "l_2",
+            "l_4",
+            "l",
+            "l_16",
+            "rgb",
+            "rgb_16_16_16",
+
+            "",
+            "la_1_1",
+            "la_2_2",
+            "la_4_4",
+            "la",
+            "la_16_16",
+            "rgba",
+            "rgba_16_16_16_16",
+        ];
+
+        int intFormat;
+        import core.bitop: bsf;
+        final switch (colorType)
+        {
+            case PngColorType.grayscale:
+                intFormat = bitDepth.bsf() + 1;
```

Comment:
this use of `bsf` is a weird way to encode the bit depth... is PNG specced such that it supports other depths? Like 3, 5, 10, 12 bpp?

Reply:
No, it just supports 1, 2, 4, 8 and 16 bits. This is actually just my arbitrary encoding with the goal to get an index into an above array based on colorType and bitDepth, and the reason the whole thing is done that way is because I was trying to avoid an allocation for format string.

Reply:
but it looks like you're decoding their depth field, which is either `0x0001` `0x0003` `0x000F` `0x00FF` or `0xFFFF`? Like, do they express depth as a mask, rather than a bit count?

Reply:
Oh, maybe I misread; I just saw `bfs` and assumed the rest... why does the bit *position* specify the format?

Reply:
Their bitDepth is really 8 or 0b1000 and I use bsf to get the index of first set bit so bsf(8)+1== 4. That way I convert their 1,2,4,8,16 to 1,2,3,4,5 which are the indices in my array for those bitDepths.

Reply:
That is just the way I decided to implement it and since it is obviously confusing I'll see about changing it.

Reply:
Ohhh, yeah i get it ;) ... yeah, that's clever, but not immediately obvious :P